diff -up gnome-shell-2.31.5/src/gnome-shell.in.jx gnome-shell-2.31.5/src/gnome-shell.in
--- gnome-shell-2.31.5/src/gnome-shell.in.jx	2010-06-25 12:47:26.000000000 -0400
+++ gnome-shell-2.31.5/src/gnome-shell.in	2010-08-12 14:44:39.000000000 -0400
@@ -213,6 +213,7 @@ def start_shell(perf_output=None):
     env = dict(os.environ)
     env.update({'GNOME_SHELL_JS'      : '@GJS_JS_DIR@:@GJS_JS_NATIVE_DIR@:' + js_dir,
                 'PATH'                : '@MUTTER_BIN_DIR@:' + os.environ.get('PATH', ''),
+		'CLUTTER_VBLANK'      : 'none',
                 'XDG_CONFIG_DIRS'     : '@sysconfdir@/xdg:' + (os.environ.get('XDG_CONFIG_DIRS') or '/etc/xdg'),
                 'GNOME_DISABLE_CRASH_DIALOG' : '1'})
 
