From b2520bd345339bbf74ab0ee4e2194206f80f4d2f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kristian=20H=C3=B8gsberg?= <krh@bitplanet.net>
Date: Mon, 16 Sep 2013 11:51:57 -0700
Subject: [PATCH 27/38] xwayland: Remove unused variables

---
 hw/xfree86/xwayland/xwayland-window.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/hw/xfree86/xwayland/xwayland-window.c b/hw/xfree86/xwayland/xwayland-window.c
index 88f57ce..20db80c 100644
--- a/hw/xfree86/xwayland/xwayland-window.c
+++ b/hw/xfree86/xwayland/xwayland-window.c
@@ -92,9 +92,6 @@ xwl_create_window(WindowPtr window)
 {
     ScreenPtr screen = window->drawable.pScreen;
     struct xwl_screen *xwl_screen;
-    char buffer[32];
-    int len, rc;
-    Atom name;
     Bool ret;
 
     xwl_screen = xwl_screen_get(screen);
-- 
1.8.4.2

