From f6f1c358958303aaa0cb2ae40044dd8cdc756618 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kristian=20H=C3=B8gsberg?= <krh@bitplanet.net>
Date: Fri, 11 Oct 2013 16:14:54 -0700
Subject: [PATCH 32/39] xwayland: Also look for wlglamor

---
 hw/xfree86/common/xf86AutoConfig.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hw/xfree86/common/xf86AutoConfig.c b/hw/xfree86/common/xf86AutoConfig.c
index b35cb96..9576b86 100644
--- a/hw/xfree86/common/xf86AutoConfig.c
+++ b/hw/xfree86/common/xf86AutoConfig.c
@@ -285,6 +285,7 @@ listPossibleVideoDrivers(char *matches[], int nmatches)
     if (i < (nmatches - 1)) {
 #ifdef XORG_WAYLAND
         if (xorgWayland)
+            matches[i++] = xnfstrdup("wlglamor");
             matches[i++] = xnfstrdup("wayland");
         else
 #endif
--
1.8.3.1
