From 408c17759542df3c906fc6fd0c01a3963ac0f345 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kristian=20H=C3=B8gsberg?= <krh@bitplanet.net>
Date: Wed, 4 Dec 2013 10:27:07 -0800
Subject: [PATCH 37/38] xwayland: Fix re-wrapping of RealizeWindow

---
 hw/xfree86/xwayland/xwayland-window.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/xfree86/xwayland/xwayland-window.c b/hw/xfree86/xwayland/xwayland-window.c
index c03855c..ee95371 100644
--- a/hw/xfree86/xwayland/xwayland-window.c
+++ b/hw/xfree86/xwayland/xwayland-window.c
@@ -110,7 +110,7 @@ xwl_realize_window(WindowPtr window)
 
     screen->RealizeWindow = xwl_screen->RealizeWindow;
     ret = (*screen->RealizeWindow)(window);
-    xwl_screen->RealizeWindow = xwl_screen->RealizeWindow;
+    xwl_screen->RealizeWindow = screen->RealizeWindow;
     screen->RealizeWindow = xwl_realize_window;
 
     if (xwl_screen->flags & XWL_FLAGS_ROOTLESS) {
-- 
1.8.4.2

