diff --git a/clutter/clutter/clutter-stage.c b/clutter/clutter/clutter-stage.c
index d0668768b079566f697571a22fc3568b90d55546..93656068b160132cdf1d439154be0782c2c58152 100644
--- a/clutter/clutter/clutter-stage.c
+++ b/clutter/clutter/clutter-stage.c
@@ -148,5 +148,7 @@ struct _ClutterStagePrivate
 
   int update_freeze_count;
 
+  int64_t last_update_time;
+
   guint redraw_pending         : 1;
   guint is_cursor_visible      : 1;
@@ -3700,15 +3702,20 @@ gint64
 _clutter_stage_get_update_time (ClutterStage *stage)
 {
   ClutterStageWindow *stage_window;
+  ClutterStagePrivate *priv = stage->priv;
 
   if (CLUTTER_ACTOR_IN_DESTRUCTION (stage))
     return 0;
 
+  if (priv->last_update_time > g_get_monotonic_time ())
+    return priv->last_update_time;
+
   stage_window = _clutter_stage_get_window (stage);
   if (stage_window == NULL)
     return 0;
 
-  return _clutter_stage_window_get_update_time (stage_window);
+  priv->last_update_time = _clutter_stage_window_get_update_time (stage_window);
+  return priv->last_update_time;
 }
 
 /**
