diff --git a/src/compositor/meta-background-actor.c b/src/compositor/meta-background-actor.c
index 466dfc2ab4689ff48fe922f4cdf4086360117517..54a6fa5f4c968b4cfba7c005b72aaf6d6424939e 100644
--- a/src/compositor/meta-background-actor.c
+++ b/src/compositor/meta-background-actor.c
@@ -193,7 +193,12 @@ set_clip_region (MetaBackgroundActor *self,
 {
   g_clear_pointer (&self->clip_region, cairo_region_destroy);
   if (clip_region)
-    self->clip_region = cairo_region_copy (clip_region);
+    {
+      if (cairo_region_is_empty (clip_region))
+        self->clip_region = cairo_region_reference (clip_region);
+      else
+        self->clip_region = cairo_region_copy (clip_region);
+    }
 }
 
 static void
@@ -202,7 +207,12 @@ set_unobscured_region (MetaBackgroundActor *self,
 {
   g_clear_pointer (&self->unobscured_region, cairo_region_destroy);
   if (unobscured_region)
-    self->unobscured_region = cairo_region_copy (unobscured_region);
+    {
+      if (cairo_region_is_empty (unobscured_region))
+        self->unobscured_region = cairo_region_reference (unobscured_region);
+      else
+        self->unobscured_region = cairo_region_copy (unobscured_region);
+    }
 }
 
 static void
diff --git a/src/compositor/meta-surface-actor.c b/src/compositor/meta-surface-actor.c
index 15bf24c07e3f426c223163265ce510eaab91f4d2..ff66f3c0aaffaab684a3c13486a79dc6a2b14e33 100644
--- a/src/compositor/meta-surface-actor.c
+++ b/src/compositor/meta-surface-actor.c
@@ -103,21 +103,28 @@ set_unobscured_region (MetaSurfaceActor *surface_actor,
   g_clear_pointer (&priv->unobscured_region, cairo_region_destroy);
   if (unobscured_region)
     {
-      cairo_rectangle_int_t bounds = { 0, };
-      float width, height;
+      if (cairo_region_is_empty (unobscured_region))
+        {
+          priv->unobscured_region = cairo_region_reference (unobscured_region);
+        }
+      else
+        {
+          cairo_rectangle_int_t bounds = { 0, };
+          float width, height;
 
-      clutter_content_get_preferred_size (CLUTTER_CONTENT (priv->texture),
-                                          &width,
-                                          &height);
-      bounds = (cairo_rectangle_int_t) {
-        .width = width,
-        .height = height,
-      };
+          clutter_content_get_preferred_size (CLUTTER_CONTENT (priv->texture),
+                                              &width,
+                                              &height);
+          bounds = (cairo_rectangle_int_t) {
+            .width = width,
+            .height = height,
+          };
 
-      priv->unobscured_region =
-        get_scaled_region (surface_actor, unobscured_region);
+          priv->unobscured_region =
+            get_scaled_region (surface_actor, unobscured_region);
 
-      cairo_region_intersect_rectangle (priv->unobscured_region, &bounds);
+          cairo_region_intersect_rectangle (priv->unobscured_region, &bounds);
+        }
     }
 }
 
@@ -130,7 +137,12 @@ set_clip_region (MetaSurfaceActor *surface_actor,
 
   g_clear_pointer (&priv->clip_region, cairo_region_destroy);
   if (clip_region)
-    priv->clip_region = get_scaled_region (surface_actor, clip_region);
+    {
+      if (cairo_region_is_empty (clip_region))
+        priv->clip_region = cairo_region_reference (clip_region);
+      else
+        priv->clip_region = get_scaled_region (surface_actor, clip_region);
+    }
 }
 
 static void
@@ -274,6 +286,18 @@ meta_surface_actor_cull_out (MetaCullable   *cullable,
 
       opaque_region = meta_shaped_texture_get_opaque_region (priv->texture);
 
+      if (!opaque_region && meta_shaped_texture_is_opaque (priv->texture))
+        {
+          cairo_rectangle_int_t rect;
+
+          rect = (cairo_rectangle_int_t) {
+            .width = meta_shaped_texture_get_width (priv->texture),
+            .height = meta_shaped_texture_get_height (priv->texture)
+          };
+
+          opaque_region = cairo_region_create_rectangle (&rect);
+        }
+
       if (!opaque_region)
         return;
 
diff --git a/src/compositor/meta-window-actor-x11.c b/src/compositor/meta-window-actor-x11.c
index 15cfa5fcf2924014bb4ebf2d2e8382dc4e830d84..ef50b1adadd20d016db98a8ac43dc94e360eb46e 100644
--- a/src/compositor/meta-window-actor-x11.c
+++ b/src/compositor/meta-window-actor-x11.c
@@ -1451,45 +1451,6 @@ meta_window_actor_x11_cull_out (MetaCullable   *cullable,
 
   cullable_parent_iface->cull_out (cullable, unobscured_region, clip_region);
 
-  if ((unobscured_region || clip_region) &&
-      meta_window_actor_is_opaque (META_WINDOW_ACTOR (self)))
-    {
-      MetaWindow *window =
-        meta_window_actor_get_meta_window (META_WINDOW_ACTOR (self));
-
-      if (window->shape_region)
-        {
-          if (unobscured_region)
-            cairo_region_subtract (unobscured_region, window->shape_region);
-          if (clip_region)
-            cairo_region_subtract (clip_region, window->shape_region);
-        }
-      else
-        {
-          cairo_region_t *region = meta_window_get_frame_bounds (window);
-
-          if (region)
-            {
-              if (unobscured_region)
-                cairo_region_subtract (unobscured_region, region);
-              if (clip_region)
-                cairo_region_subtract (clip_region, region);
-            }
-          else
-            {
-              cairo_rectangle_int_t rect;
-
-              meta_window_get_frame_rect (window, &rect);
-              rect.x = rect.y = 0;
-
-              if (unobscured_region)
-                cairo_region_subtract_rectangle (unobscured_region, &rect);
-              if (clip_region)
-                cairo_region_subtract_rectangle (clip_region, &rect);
-            }
-        }
-    }
-
   set_clip_region_beneath (self, clip_region);
 }
 
