diff --git a/src/backends/native/meta-cursor-renderer-native.c b/src/backends/native/meta-cursor-renderer-native.c
index 03f18e8b6ae7d684b43243c40387f3d7ff7bf2e5..8c6b424d0cd0c5b8c9dacc7e50996e43044d6861 100644
--- a/src/backends/native/meta-cursor-renderer-native.c
+++ b/src/backends/native/meta-cursor-renderer-native.c
@@ -1331,7 +1331,7 @@ realize_cursor_sprite_from_wl_buffer_for_gpu (MetaCursorRenderer      *renderer,
       buffer_data = wl_shm_buffer_get_data (shm_buffer);
 
       width = wl_shm_buffer_get_width (shm_buffer);
-      height = wl_shm_buffer_get_width (shm_buffer);
+      height = wl_shm_buffer_get_height (shm_buffer);
 
       switch (wl_shm_buffer_get_format (shm_buffer))
         {
