diff --git a/clutter/clutter/cogl/clutter-stage-cogl.c b/clutter/clutter/cogl/clutter-stage-cogl.c
index db87e39337c39ce54394e0caea75fc44b968c7df..f7f71e6e169b0c41e5bd6d4d8176083bd1daa938 100644
--- a/clutter/clutter/cogl/clutter-stage-cogl.c
+++ b/clutter/clutter/cogl/clutter-stage-cogl.c
@@ -551,7 +551,7 @@ clutter_stage_cogl_redraw_view (ClutterStageWindow *stage_window,
   CoglFramebuffer *onscreen = clutter_stage_view_get_onscreen (view);
   cairo_rectangle_int_t view_rect;
   gboolean is_full_redraw;
-  gboolean use_clipped_redraw;
+  gboolean use_clipped_redraw = TRUE;
   gboolean can_blit_sub_buffer;
   gboolean has_buffer_age;
   gboolean swap_with_damage;
@@ -561,7 +561,7 @@ clutter_stage_cogl_redraw_view (ClutterStageWindow *stage_window,
   cairo_region_t *swap_region;
   float fb_scale;
   int fb_width, fb_height;
-  int buffer_age;
+  int buffer_age = 0;
   gboolean res;
 
   clutter_stage_view_get_layout (view, &view_rect);
