diff --git a/src/st/st-label.c b/src/st/st-label.c
index e09b1b0c7cb71cf87c75e2cd2b686269de2b4f7f..7415dec29ceff071b656ec3303073d81928e0cbe 100644
--- a/src/st/st-label.c
+++ b/src/st/st-label.c
@@ -293,6 +293,7 @@ st_label_class_init (StLabelClass *klass)
 static void
 st_label_init (StLabel *label)
 {
+  ClutterActor *actor = CLUTTER_ACTOR (label);
   StLabelPrivate *priv;
 
   label->priv = priv = st_label_get_instance_private (label);
@@ -304,7 +305,10 @@ st_label_init (StLabel *label)
   label->priv->shadow_width = -1.;
   label->priv->shadow_height = -1.;
 
-  clutter_actor_add_child (CLUTTER_ACTOR (label), priv->label);
+  clutter_actor_add_child (actor, priv->label);
+
+  clutter_actor_set_offscreen_redirect (actor,
+                                        CLUTTER_OFFSCREEN_REDIRECT_ALWAYS);
 }
 
 /**
