diff --git a/src/compositor/meta-background-actor.c b/src/compositor/meta-background-actor.c
index 361fbeeb1fa69010256e171fc7d7943ba2b4e201..628fb3ff3ba3a86897a75029d2d18e3160937d27 100644
--- a/src/compositor/meta-background-actor.c
+++ b/src/compositor/meta-background-actor.c
@@ -354,7 +354,7 @@ setup_pipeline (MetaBackgroundActor   *self,
   guint8 opacity;
   float color_component;
   CoglFramebuffer *fb;
-  CoglPipelineFilter filter;
+  CoglPipelineFilter min_filter, mag_filter;
 
   opacity = clutter_actor_get_paint_opacity (CLUTTER_ACTOR (self));
   if (opacity < 255)
@@ -455,11 +455,17 @@ setup_pipeline (MetaBackgroundActor   *self,
                                          actor_pixel_rect->width,
                                          actor_pixel_rect->height,
                                          NULL, NULL))
-    filter = COGL_PIPELINE_FILTER_NEAREST;
+    {
+      min_filter = COGL_PIPELINE_FILTER_NEAREST;
+      mag_filter = COGL_PIPELINE_FILTER_NEAREST;
+    }
   else
-    filter = COGL_PIPELINE_FILTER_LINEAR;
+    {
+      min_filter = COGL_PIPELINE_FILTER_LINEAR_MIPMAP_NEAREST;
+      mag_filter = COGL_PIPELINE_FILTER_LINEAR;
+    }
 
-  cogl_pipeline_set_layer_filters (self->pipeline, 0, filter, filter);
+  cogl_pipeline_set_layer_filters (self->pipeline, 0, min_filter, mag_filter);
 }
 
 static void
