diff --git a/src/backends/meta-backend-private.h b/src/backends/meta-backend-private.h
index 6d7d3ad7587a4959f46e44d95ebce0eca69de933..5f4981fda55d08cbebdd179d1ca14a928721826d 100644
--- a/src/backends/meta-backend-private.h
+++ b/src/backends/meta-backend-private.h
@@ -42,10 +42,6 @@
 #include "backends/meta-settings-private.h"
 #include "core/util-private.h"
 
-#ifdef HAVE_REMOTE_DESKTOP
-#include "backends/meta-remote-desktop.h"
-#endif
-
 #define DEFAULT_XKB_RULES_FILE "evdev"
 #define DEFAULT_XKB_MODEL "pc105+inet"
 
diff --git a/src/backends/meta-backend-types.h b/src/backends/meta-backend-types.h
index 4753c07440db8d419ef445aa5a67bd956bd7524c..146a8c3d7c3b1e6e38dce033342be5b8173bded0 100644
--- a/src/backends/meta-backend-types.h
+++ b/src/backends/meta-backend-types.h
@@ -59,4 +59,8 @@ typedef struct _MetaScreenCastStream MetaScreenCastStream;
 
 typedef struct _MetaWaylandCompositor MetaWaylandCompositor;
 
+#ifdef HAVE_REMOTE_DESKTOP
+typedef struct _MetaRemoteDesktop MetaRemoteDesktop;
+#endif
+
 #endif /* META_BACKEND_TYPE_H */
