diff --git a/clutter/clutter/clutter-stage.c b/clutter/clutter/clutter-stage.c
index c8477972efb361d26977e2ca251b8417809c6941..fd2d3008c449529574bbc458e6efe6517ebff6ea 100644
--- a/clutter/clutter/clutter-stage.c
+++ b/clutter/clutter/clutter-stage.c
@@ -3049,7 +3049,7 @@ clutter_stage_set_key_focus (ClutterStage *stage,
   if (actor != NULL)
     {
       priv->key_focused_actor = actor;
-      _clutter_actor_set_has_key_focus (actor, FALSE);
+      _clutter_actor_set_has_key_focus (actor, TRUE);
     }
   else
     _clutter_actor_set_has_key_focus (CLUTTER_ACTOR (stage), TRUE);
