diff --git a/clutter/clutter/cogl/clutter-stage-cogl.c b/clutter/clutter/cogl/clutter-stage-cogl.c
index 3d0af33484c4eb9c0fdaddcb007f6c906d23bc6a..9c8f91298abbef44f59bfc4855b9615b8e608c84 100644
--- a/clutter/clutter/cogl/clutter-stage-cogl.c
+++ b/clutter/clutter/cogl/clutter-stage-cogl.c
@@ -771,8 +771,9 @@ clutter_stage_cogl_redraw_view (ClutterStageWindow *stage_window,
     }
   else
     {
-      fb_clip_region = cairo_region_create ();
-      redraw_clip = cairo_region_reference (fb_clip_region);
+      cairo_rectangle_int_t rect = { 0, 0, view_rect.width, view_rect.height };
+      fb_clip_region = cairo_region_create_rectangle (&rect);
+      redraw_clip = cairo_region_copy (fb_clip_region);
     }
 
   if (may_use_clipped_redraw &&
@@ -786,7 +787,7 @@ clutter_stage_cogl_redraw_view (ClutterStageWindow *stage_window,
   swap_with_damage = FALSE;
   if (has_buffer_age)
     {
-      if (use_clipped_redraw && !clip_region_empty)
+      if (use_clipped_redraw)
         {
           int age;
 
