diff --git a/js/ui/iconGrid.js b/js/ui/iconGrid.js
index 67691c7d50a6202898add543eedc8a0e28413911..fb307734f75d3b255c2dc071cecb1c135abe663a 100644
--- a/js/ui/iconGrid.js
+++ b/js/ui/iconGrid.js
@@ -274,23 +274,11 @@ var IconGrid = GObject.registerClass({
 
     _childAdded(grid, child) {
         child._iconGridKeyFocusInId = child.connect('key-focus-in', this._keyFocusIn.bind(this));
-
-        child._paintVisible = child.opacity > 0;
-        child._opacityChangedId = child.connect('notify::opacity', () => {
-            let paintVisible = child._paintVisible;
-            child._paintVisible = child.opacity > 0;
-            if (paintVisible !== child._paintVisible)
-                this.queue_relayout();
-        });
     }
 
     _childRemoved(grid, child) {
         child.disconnect(child._iconGridKeyFocusInId);
         delete child._iconGridKeyFocusInId;
-
-        child.disconnect(child._opacityChangedId);
-        delete child._opacityChangedId;
-        delete child._paintVisible;
     }
 
     vfunc_get_preferred_width(_forHeight) {
@@ -463,6 +451,8 @@ var IconGrid = GObject.registerClass({
     _resetAnimationActors() {
         this._clonesAnimating.forEach(clone => {
             clone.source.reactive = true;
+            if (clone.source.opacity === 0)
+                clone.source.queue_relayout();  // gnome-shell#1502
             clone.source.opacity = 255;
             clone.destroy();
         });
