diff --git a/src/compositor/meta-window-actor.c b/src/compositor/meta-window-actor.c
index a42dfaf7b8c1146e667078e19b224315b595f4fc..3ebfc81e6ca13d7d7a3d156b6cd6e2bc72452671 100644
--- a/src/compositor/meta-window-actor.c
+++ b/src/compositor/meta-window-actor.c
@@ -957,20 +957,28 @@ meta_window_actor_cull_out (MetaCullable   *cullable,
 
   meta_cullable_cull_out_children (cullable, unobscured_region, clip_region);
 
-  if (unobscured_region && meta_window_actor_is_opaque (self))
+  if ((unobscured_region || clip_region) && meta_window_actor_is_opaque (self))
     {
       cairo_region_t *region = meta_window_get_frame_bounds (priv->window);
 
       if (region)
         {
-          cairo_region_subtract (unobscured_region, region);
+          if (unobscured_region)
+            cairo_region_subtract (unobscured_region, region);
+          if (clip_region)
+            cairo_region_subtract (clip_region, region);
         }
       else
         {
           cairo_rectangle_int_t rect;
+
           meta_window_get_frame_rect (priv->window, &rect);
           rect.x = rect.y = 0;
-          cairo_region_subtract_rectangle (unobscured_region, &rect);
+
+          if (unobscured_region)
+            cairo_region_subtract_rectangle (unobscured_region, &rect);
+          if (clip_region)
+            cairo_region_subtract_rectangle (clip_region, &rect);
         }
     }
 }
