From 2c53e4d6fcb16784548bd755827de1ee5ddddeb8 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Thu, 8 Nov 2007 15:33:08 -0500
Subject: [PATCH] Speed up DDC probe.

---
 hw/xfree86/ddc/xf86DDC.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/hw/xfree86/ddc/xf86DDC.c b/hw/xfree86/ddc/xf86DDC.c
index e47b8b8..9d38aab 100644
--- a/hw/xfree86/ddc/xf86DDC.c
+++ b/hw/xfree86/ddc/xf86DDC.c
@@ -148,6 +148,9 @@ xf86DoEDID_DDC2(int scrnIndex, I2CBusPtr pBus)
     if (noddc || noddc2)
 	return NULL;
 
+    if (!xf86I2CProbeAddress(pBus, 0xA1))
+	return NULL;
+
     EDID_block = EDID1Read_DDC2(scrnIndex,pBus);
 
     if (EDID_block){
-- 
1.5.3.4

