diff -up gnome-shell-3.5.3/src/shell-app-system.c.avoid-redhat-menus gnome-shell-3.5.3/src/shell-app-system.c
--- gnome-shell-3.5.3/src/shell-app-system.c.avoid-redhat-menus	2012-06-25 12:40:06.000000000 -0400
+++ gnome-shell-3.5.3/src/shell-app-system.c	2012-06-26 00:45:55.824232085 -0400
@@ -103,7 +103,7 @@ shell_app_system_init (ShellAppSystem *s
 
   /* We want to track NoDisplay apps, so we add INCLUDE_NODISPLAY. We'll
    * filter NoDisplay apps out when showing them to the user. */
-  priv->apps_tree = gmenu_tree_new ("applications.menu", GMENU_TREE_FLAGS_INCLUDE_NODISPLAY);
+  priv->apps_tree = gmenu_tree_new ("applications-gnome.menu", GMENU_TREE_FLAGS_INCLUDE_NODISPLAY);
   g_signal_connect (priv->apps_tree, "changed", G_CALLBACK (on_apps_tree_changed_cb), self);
 
   priv->settings_tree = gmenu_tree_new ("gnomecc.menu", 0);
