diff -up gnome-shell-3.3.3/configure.ac.gl gnome-shell-3.3.3/configure.ac
--- gnome-shell-3.3.3/configure.ac.gl	2012-01-19 21:58:03.637964155 -0500
+++ gnome-shell-3.3.3/configure.ac	2012-01-19 21:59:19.634969153 -0500
@@ -84,6 +84,7 @@ PKG_CHECK_MODULES(GNOME_SHELL, gio-unix-
                                gjs-internals-1.0 >= $GJS_MIN_VERSION
 			       libgnome-menu-3.0 $recorder_modules
                                gdk-x11-3.0 libsoup-2.4
+                               gl
 			       clutter-x11-1.0 >= $CLUTTER_MIN_VERSION
 			       clutter-glx-1.0 >= $CLUTTER_MIN_VERSION
                                libstartup-notification-1.0 >= $STARTUP_NOTIFICATION_MIN_VERSION
