From 82aea3e8d6fc2a16759fd6f232c7358989a5f607 Mon Sep 17 00:00:00 2001
From: Alexander Larsson <alexl@redhat.com>
Date: Fri, 20 May 2011 11:14:54 +0200
Subject: [PATCH] Import Shell as needed by recent commit

---
 js/ui/windowManager.js |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/js/ui/windowManager.js b/js/ui/windowManager.js
index 52a00da..a3e2918 100644
--- a/js/ui/windowManager.js
+++ b/js/ui/windowManager.js
@@ -6,6 +6,7 @@ const Gio = imports.gi.Gio;
 const Lang = imports.lang;
 const Meta = imports.gi.Meta;
 const St = imports.gi.St;
+const Shell = imports.gi.Shell;
 
 const AltTab = imports.ui.altTab;
 const WorkspaceSwitcherPopup = imports.ui.workspaceSwitcherPopup;
-- 
1.7.5.1

