diff -up mutter-2.28.1/src/tools/Makefile.am.jx mutter-2.28.1/src/tools/Makefile.am
--- mutter-2.28.1/src/tools/Makefile.am.jx	2010-01-25 11:07:32.000000000 -0500
+++ mutter-2.28.1/src/tools/Makefile.am	2010-02-16 18:59:50.000000000 -0500
@@ -26,7 +26,7 @@ noinst_PROGRAMS=mutter-mag mutter-graysc
 
 mutter_message_LDADD= @MUTTER_MESSAGE_LIBS@
 mutter_window_demo_LDADD= @MUTTER_WINDOW_DEMO_LIBS@
-mutter_mag_LDADD= @MUTTER_WINDOW_DEMO_LIBS@
+mutter_mag_LDADD= @MUTTER_WINDOW_DEMO_LIBS@ -lm
 mutter_grayscale_LDADD = @MUTTER_WINDOW_DEMO_LIBS@
 
 EXTRA_DIST=$(icon_DATA)
