From a1ac1f4b2ec41e6da572a29ff5a1c20bd0e6934c Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Wed, 29 Jul 2009 16:08:53 +0000
Subject: Build using Clutter 1.0

Clutter just released 1.0, sync to that.
---
diff --git a/src/Makefile.am b/src/Makefile.am
index b675ab8..2ee9978 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -210,8 +210,8 @@ Meta-$(api_version).gir: $(G_IR_SCANNER) mutter $(libmutterinclude_HEADERS) $(mu
 		--include=GObject-2.0			\
 		--include=Gdk-2.0			\
 		--include=Gtk-2.0			\
-		--include=Clutter-0.9			\
-		--pkg=clutter-0.9			\
+		--include=Clutter-1.0			\
+		--pkg=clutter-1.0			\
 		--pkg=gtk+-2.0				\
 		--include=xfixes-4.0			\
 		--program=$$pwd/mutter			\
--
cgit v0.8.2
