diff --git a/configure b/configure
index 5193efb..9ae5f3d 100755
--- a/configure
+++ b/configure
@@ -16600,7 +16600,7 @@ fi
 
 if test "$enable_remote_desktop" = "yes"; then :
 
-  MUTTER_PC_MODULES="$MUTTER_PC_MODULES libpipewire-0.1 >= 0.2.0"
+  MUTTER_PC_MODULES="$MUTTER_PC_MODULES libpipewire-0.2 >= 0.2.2"
 
 $as_echo "#define HAVE_REMOTE_DESKTOP 1" >>confdefs.h
 
diff --git a/configure.ac b/configure.ac
index d91f120..a404fa2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -244,7 +244,7 @@ AC_ARG_ENABLE(remote-desktop,
   enable_remote_desktop=no
 )
 AS_IF([test "$enable_remote_desktop" = "yes"], [
-  MUTTER_PC_MODULES="$MUTTER_PC_MODULES libpipewire-0.1 >= 0.2.0"
+  MUTTER_PC_MODULES="$MUTTER_PC_MODULES libpipewire-0.2 >= 0.2.2"
   AC_DEFINE([HAVE_REMOTE_DESKTOP],[1], [Defined if screen cast and remote desktop support is enabled])
 ])
 AM_CONDITIONAL([HAVE_REMOTE_DESKTOP],[test "$enable_remote_desktop" = "yes"])
