--- xorg-server-1.1.1/include/dix-config.h.in.builderstring	2006-07-05 14:38:48.000000000 -0400
+++ xorg-server-1.1.1/include/dix-config.h.in	2006-08-09 16:59:05.000000000 -0400
@@ -15,6 +15,9 @@
 /* Operating System Vendor */
 #undef OSVENDOR
 
+/* Builder string */
+#undef BUILDERSTRING
+
 /* Default font path */
 #undef COMPILEDDEFAULTFONTPATH
 
--- xorg-server-1.1.1/configure.ac.builderstring	2006-08-09 14:04:19.000000000 -0400
+++ xorg-server-1.1.1/configure.ac	2006-08-09 16:58:25.000000000 -0400
@@ -358,6 +358,9 @@
 AC_ARG_WITH(os-vendor,       AS_HELP_STRING([--with-os-vendor=OSVENDOR], [Name of OS vendor]),
 				[ OSVENDOR="$withval" ],
 				[ OSVENDOR="" ])
+AC_ARG_WITH(builderstring,   AS_HELP_STRING([--with-builderstring=BUILDERSTRING], [Additional builder string]),
+				[ BUILDERSTRING="$withval" ]
+				[ ])
 AC_ARG_WITH(mesa-source,     AS_HELP_STRING([--with-mesa-source=MESA_SOURCE], [Path to Mesa source tree]),
                                 [ MESA_SOURCE="$withval" ],
                                 [ MESA_SOURCE="" ])
@@ -817,6 +820,8 @@
 AC_DEFINE_UNQUOTED(BUILDERADDR, ["$BUILDERADDR"], [Builder address])
 AC_DEFINE_UNQUOTED(OSNAME, ["$OSNAME"], [Operating System Name])
 AC_DEFINE_UNQUOTED(OSVENDOR, ["$OSVENDOR"], [Operating System Vendor])
+AC_DEFINE_UNQUOTED(BUILDERSTRING, ["$BUILDERSTRING"], [Builder string])
+
 AC_SUBST([VENDOR_STRING])
 AC_SUBST([VENDOR_STRING_SHORT])
 AC_SUBST([VENDOR_RELEASE])
