--- xorg-server-1.1.1/mi/miinitext.c.xkb-in-xnest	2006-07-05 14:38:48.000000000 -0400
+++ xorg-server-1.1.1/mi/miinitext.c	2006-08-17 22:57:27.000000000 -0400
@@ -595,7 +595,7 @@
 #ifdef XSYNC
     if (!noSyncExtension) SyncExtensionInit();
 #endif
-#if defined(XKB) && !defined(PRINT_ONLY_SERVER) && !defined(NO_HW_ONLY_EXTS)
+#if defined(XKB) && !defined(PRINT_ONLY_SERVER)
     if (!noXkbExtension) XkbExtensionInit();
 #endif
 #ifdef XCMISC
