From ad677238bc96a8578113bbe76d605d7a87aca44c Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 14 Oct 2008 13:00:50 -0400
Subject: [PATCH] Add backtrace definitions to dix-config.h.in

---
 include/dix-config.h.in |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/include/dix-config.h.in b/include/dix-config.h.in
index 26f4b6a..6c3d91c 100644
--- a/include/dix-config.h.in
+++ b/include/dix-config.h.in
@@ -75,6 +75,9 @@
 /* Define to 1 if you have the <asm/mtrr.h> header file. */
 #undef HAVE_ASM_MTRR_H
 
+/* Has backtrace support */
+#undef HAVE_BACKTRACE
+
 /* Define to 1 if you have the <byteswap.h> header file. */
 #undef HAVE_BYTESWAP_H
 
@@ -94,6 +97,9 @@
 /* Define to 1 if you don't have `vprintf' but do have `_doprnt.' */
 #undef HAVE_DOPRNT
 
+/* Have execinfo.h */
+#undef HAVE_EXECINFO_H
+
 /* Define to 1 if you have the <fcntl.h> header file. */
 #undef HAVE_FCNTL_H
 
-- 
1.6.0.1

