From 6f3a12d390eb67856f2a32a83f72eed3aea99e70 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Thu, 27 Mar 2008 09:50:34 -0400
Subject: [PATCH] Use vmmouse, not mouse, for automatic mouse sections.

---
 hw/xfree86/common/xf86Config.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/hw/xfree86/common/xf86Config.c b/hw/xfree86/common/xf86Config.c
index 4a4aabc..0c579fb 100644
--- a/hw/xfree86/common/xf86Config.c
+++ b/hw/xfree86/common/xf86Config.c
@@ -1336,7 +1336,11 @@ checkCoreInputDevices(serverLayoutPtr servlayoutp, Bool implicitLayout)
     if (!foundPointer) {
 	bzero(&defPtr, sizeof(defPtr));
 	defPtr.inp_identifier = "<default pointer>";
+#if defined(__i386__) || defined (__amd64__)
+	defPtr.inp_driver = "vmmouse"
+#else
 	defPtr.inp_driver = "mouse";
+#endif
 	confInput = &defPtr;
 	foundPointer = TRUE;
 	from = X_DEFAULT;
@@ -1383,7 +1387,11 @@ checkCoreInputDevices(serverLayoutPtr servlayoutp, Bool implicitLayout)
 	xf86Msg(X_INFO, "No default mouse found, adding one\n");
 	bzero(&defPtr, sizeof(defPtr));
 	defPtr.inp_identifier = "<default pointer>";
+#if defined(__i386__) || defined (__amd64__)
+	defPtr.inp_driver = "vmmouse"
+#else
 	defPtr.inp_driver = "mouse";
+#endif
 	confInput = &defPtr;
 	foundPointer = configInput(&Pointer, confInput, from);
         if (foundPointer) {
-- 
1.5.4.3

