From 9ee22b4f743a2850fed35f179c56f94ce4bb5242 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 3 Mar 2009 10:58:33 -0500
Subject: [PATCH] Primary video device hack

---
 hw/xfree86/common/xf86pciBus.c |   54 ++++++++++++++++++++++++++++++++++-----
 1 files changed, 47 insertions(+), 7 deletions(-)

diff --git a/hw/xfree86/common/xf86pciBus.c b/hw/xfree86/common/xf86pciBus.c
index 467a0c3..f9999d8 100644
--- a/hw/xfree86/common/xf86pciBus.c
+++ b/hw/xfree86/common/xf86pciBus.c
@@ -341,6 +341,39 @@ restorePciBusState(BusAccPtr ptr)
 }
 #undef MASKBITS
 
+/* oh god what have i done */
+static Bool
+looks_like_bios_primary(struct pci_device *info)
+{
+    unsigned char *bios;
+    unsigned short vendor, device;
+    int offset;
+    Bool ret = FALSE;
+
+    bios = xf86MapVidMem(-1, VIDMEM_MMIO, 0xc0000, 0x10000);
+    if (!bios)
+        return FALSE;
+
+    if (bios[0] != 0x55 || bios[1] != 0xAA)
+        goto out;
+
+    offset = (bios[0x19] << 8) + bios[0x18];
+
+    if (bios[offset] != 'P' ||
+        bios[offset+1] != 'C' ||
+        bios[offset+2] != 'I' ||
+        bios[offset+3] != 'R')
+        goto out;
+
+    vendor = (bios[offset+5] << 8) + bios[offset+4];
+    device = (bios[offset+7] << 8) + bios[offset+6];
+
+    ret = (info->vendor_id == vendor) && (info->device_id == device);
+
+out:
+    xf86UnMapVidMem(-1, bios, 0x10000);
+    return ret;
+}
 
 /*
  * xf86Bus.c interface
@@ -375,24 +408,31 @@ xf86PciProbe(void)
 	}
     }
 
-
     /* If we haven't found a primary device try a different heuristic */
     if (primaryBus.type == BUS_NONE && num) {
 	for (i = 0; i < num; i++) {
 	    uint16_t  command;
 
 	    info = xf86PciVideoInfo[i];
+            if (!IS_VGA(info->device_class))
+                continue;
+
 	    pci_device_cfg_read_u16(info, & command, 4);
 
-	    if ((command & PCI_CMD_MEM_ENABLE) 
-		&& ((num == 1) || IS_VGA(info->device_class))) {
-		if (primaryBus.type == BUS_NONE) {
+	    if ((command & PCI_CMD_MEM_ENABLE)) {
+                if (num == 1) {
 		    primaryBus.type = BUS_PCI;
 		    primaryBus.id.pci = info;
-		} else {
-		    xf86Msg(X_NOTICE,
+                    break;
+                } else if (looks_like_bios_primary(info)) {
+                    if (primaryBus.type == BUS_NONE) {
+                        primaryBus.type = BUS_PCI;
+                        primaryBus.id.pci = info;
+                    } else {
+		        xf86Msg(X_NOTICE,
 			    "More than one possible primary device found\n");
-		    primaryBus.type ^= (BusType)(-1);
+		        primaryBus.type ^= (BusType)(-1);
+                    }
 		}
 	    }
 	}
-- 
1.6.1.3

