From 59681e08381f8ca83263a6313b43905909e68583 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Wed, 28 Nov 2007 12:09:16 -0500
Subject: [PATCH] via -> openchrome

---
 hw/xfree86/common/xf86AutoConfig.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hw/xfree86/common/xf86AutoConfig.c b/hw/xfree86/common/xf86AutoConfig.c
index c6e1972..da6c3f3 100644
--- a/hw/xfree86/common/xf86AutoConfig.c
+++ b/hw/xfree86/common/xf86AutoConfig.c
@@ -184,7 +184,7 @@ videoPtrToDriverName(struct pci_device *dev)
 	case 0x3d3d:		    return "glint";
 	case 0x1023:		    return "trident";
 	case 0x100c:		    return "tseng";
-	case 0x1106:		    return "via";
+	case 0x1106:		    return "openchrome";
 	case 0x15ad:		    return "vmware";
 	default: break;
     }
-- 
1.5.3.4

