From 7d97c19c2d530c8db16b01de9b168dd910059b4e Mon Sep 17 00:00:00 2001
From: Xiang, Haihao <haihao.xiang@intel.com>
Date: Wed, 2 Apr 2008 16:29:30 +1000
Subject: [PATCH] dri2: fix crasher if DRI2Connect fails

---
 hw/xfree86/dri2/dri2ext.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hw/xfree86/dri2/dri2ext.c b/hw/xfree86/dri2/dri2ext.c
index ca2e029..d1d52a4 100644
--- a/hw/xfree86/dri2/dri2ext.c
+++ b/hw/xfree86/dri2/dri2ext.c
@@ -114,7 +114,7 @@ ProcDRI2Connect(ClientPtr client)
     ScreenPtr pScreen;
     int fd;
     const char *driverName;
-    char *busId;
+    char *busId = NULL;
     unsigned int sareaHandle;
 
     REQUEST_SIZE_MATCH(xDRI2ConnectReq);
-- 
1.5.4.5

