From db7547ab7f43f36f2eb38bc9ea29cbfa691ed7f0 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Thu, 4 Mar 2010 15:18:38 +1000
Subject: [PATCH] tweak default mode list yet again

---
 hw/xfree86/modes/xf86Crtc.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/hw/xfree86/modes/xf86Crtc.c b/hw/xfree86/modes/xf86Crtc.c
index 5875731..adae031 100644
--- a/hw/xfree86/modes/xf86Crtc.c
+++ b/hw/xfree86/modes/xf86Crtc.c
@@ -1621,6 +1621,9 @@ xf86ProbeOutputModes (ScrnInfoPtr scrn, int maxX, int maxY)
 	}
 	
 	output_modes = (*output->funcs->get_modes) (output);
+
+	/* if the driver found modes, don't second-guess it */
+	add_default_modes = (output_modes == NULL);
 	
 	edid_monitor = output->MonInfo;
 	
-- 
1.7.3.2

