From cc8ff24ba47acbe13872176086b6d75943508e61 Mon Sep 17 00:00:00 2001
From: Fedora X Ninjas <x@fedoraproject.org>
Date: Mon, 16 Feb 2009 15:50:14 +1000
Subject: [PATCH] autoconfig: select nouveau by default for NVIDIA GPUs

---
 hw/xfree86/common/xf86AutoConfig.c |   14 +++++++++++++-
 1 files changed, 13 insertions(+), 1 deletions(-)

diff --git a/hw/xfree86/common/xf86AutoConfig.c b/hw/xfree86/common/xf86AutoConfig.c
index 19c06c8..853d661 100644
--- a/hw/xfree86/common/xf86AutoConfig.c
+++ b/hw/xfree86/common/xf86AutoConfig.c
@@ -181,7 +181,19 @@ videoPtrToDriverList(struct pci_device *dev,
 	case 0x102b:		    driverList[0] = "mga";	break;
 	case 0x10c8:		    driverList[0] = "neomagic"; break;
 	case 0x105d:		    driverList[0] = "i128";	break;
-	case 0x10de: case 0x12d2:   driverList[0] = "nv";	break;
+	case 0x10de:
+	    switch (dev->device_id & 0xfff0) {
+	    /* Non-functional with both nouveau and nv */
+	    case 0x0840:
+	    case 0x0860:
+		driverList[0] = "vesa";
+		break;
+	    default:
+		driverList[0] = "nouveau";
+		break;
+	    }
+	    break;
+	case 0x12d2:		    driverList[0] = "nv";	break;
 	case 0x1163:		    driverList[0] = "rendition"; break;
 	case 0x5333:
 	    switch (dev->device_id)
-- 
1.6.1.3

