--- xorg-server-1.2.0/hw/kdrive/ephyr/hostx.c.xephyr-keysym	2007-01-22 22:13:14.000000000 -0500
+++ xorg-server-1.2.0/hw/kdrive/ephyr/hostx.c	2007-01-29 17:58:00.000000000 -0500
@@ -77,9 +77,10 @@
 
 static int            HostXWantDamageDebug = 0;
 
-extern KeySym         EphyrKeymap[];
+/* these two need to be 32-bits. do not say KeySym, that way lies madness */
+extern unsigned int   EphyrKeymap[];
+extern unsigned int   kdKeymap[];
 
-extern KeySym	      kdKeymap[];
 extern int	      kdMinScanCode;
 extern int	      kdMaxScanCode;
 extern int	      kdMinKeyCode;
@@ -662,7 +663,7 @@
 void
 hostx_load_keymap(void)
 {
-  KeySym          *keymap;
+  unsigned int     *keymap;
   int              host_width, min_keycode, max_keycode, width;
   int              i,j;
 
