From 7007f29215604b2d16c4bda25ec2b752ca05d75f Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Wed, 9 Apr 2008 13:55:25 -0400
Subject: [PATCH] Default X-SELinux to off.

---
 os/utils.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/os/utils.c b/os/utils.c
index d785d46..f58c763 100644
--- a/os/utils.c
+++ b/os/utils.c
@@ -233,7 +233,7 @@ _X_EXPORT Bool noXInputExtension = FALSE;
 _X_EXPORT Bool noXIdleExtension = FALSE;
 #endif
 #ifdef XSELINUX
-_X_EXPORT Bool noSELinuxExtension = FALSE;
+_X_EXPORT Bool noSELinuxExtension = TRUE;
 _X_EXPORT int selinuxEnforcingState = SELINUX_MODE_DEFAULT;
 #endif
 #ifdef XV
-- 
1.5.4.5

