From ce4ce1c26d9b7db94e36c72523ef2b05f0716287 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 7 Oct 2008 13:45:58 -0400
Subject: [PATCH] Plug int10 memory leak and shut up useless warnings.

---
 hw/xfree86/int10/helper_exec.c |   14 +-------------
 1 files changed, 1 insertions(+), 13 deletions(-)

diff --git a/hw/xfree86/int10/helper_exec.c b/hw/xfree86/int10/helper_exec.c
index c3af5bc..fa1516b 100644
--- a/hw/xfree86/int10/helper_exec.c
+++ b/hw/xfree86/int10/helper_exec.c
@@ -480,20 +480,8 @@ pci_device_for_cfg_address (CARD32 addr)
 	    pci_slot_match_iterator_create (&slot_match);
 	if (iter)
 		dev = pci_device_next(iter);
-	if (!dev) {
-		char buf[128]; /* enough to store "%u@%u" */
-		xf86FormatPciBusNumber(tag >> 16, buf);
-		ErrorF("Failed to find device matching %s:%u:%u\n",
-				buf, slot_match.dev, slot_match.func);
-		return NULL;
-	}
 
-	if (pci_device_next(iter)) {
-		char buf[128]; /* enough to store "%u@%u" */
-		xf86FormatPciBusNumber(tag >> 16, buf);
-		ErrorF("Multiple devices matching %s:%u:%u\n",
-				buf, slot_match.dev, slot_match.func);
-	}
+	pci_iterator_destroy(iter);
 
 	return dev;
 }
-- 
1.6.0.1

