commit 1b94c117e0f294ef2f89bf24d45ba7a8e45efe35
Author: Matthias Hopf <mhopf@suse.de>
Date:   Tue Oct 10 19:37:22 2006 +0200

    Fixing probably long-standing bug in domain support.

diff --git a/hw/xfree86/os-support/bus/linuxPci.c b/hw/xfree86/os-support/bus/linuxPci.c
index c6dad1f..7d220d0 100644
--- a/hw/xfree86/os-support/bus/linuxPci.c
+++ b/hw/xfree86/os-support/bus/linuxPci.c
@@ -498,7 +498,7 @@ xf86GetPciDomain(PCITAG Tag)
     pPCI = xf86GetPciHostConfigFromTag(Tag);
 
     if (pPCI && (result = PCI_DOM_FROM_BUS(pPCI->busnum)))
-	return result;
+	return result + 1;
 
     if (!pPCI || pPCI->fakeDevice)
 	return 1;		/* Domain 0 is reserved */
