--- xorg-server-1.2.99.905/hw/xfree86/os-support/bus/linuxPci.c.jx	2007-04-09 16:15:43.000000000 -0400
+++ xorg-server-1.2.99.905/hw/xfree86/os-support/bus/linuxPci.c	2007-04-09 17:37:28.000000000 -0400
@@ -778,8 +778,8 @@
     bus  = PCI_BUS_NO_DOMAIN(PCI_BUS_FROM_TAG(Tag));
     dev  = PCI_DEV_FROM_TAG(Tag);
     func = PCI_FUNC_FROM_TAG(Tag);
-    sprintf(file, "/sys/devices/pci%04x:%02x/%04x:%02x:%02x.%1x/rom",
-	    dom, bus, dom, bus, dev, func);
+    sprintf(file, "/sys/bus/pci/devices/%04x:%02x:%02x.%1x/rom",
+	    dom, bus, dev, func);
 
     /*
      * If the caller wants the ROM and the sysfs rom interface exists,
