From: Adam Jackson <ajax@redhat.com>
Date: Sun, 28 Oct 2007 09:37:52 +0100
Subject: [PATCH] Document default font path correctly.

---
diff -up xorg-server-1.3.0.0/hw/xfree86/doc/man/xorg.conf.man.pre.jx xorg-server-1.3.0.0/hw/xfree86/doc/man/xorg.conf.man.pre
--- xorg-server-1.3.0.0/hw/xfree86/doc/man/xorg.conf.man.pre.jx	2006-12-22 03:06:49.000000000 -0500
+++ xorg-server-1.3.0.0/hw/xfree86/doc/man/xorg.conf.man.pre	2007-08-23 15:27:47.000000000 -0400
@@ -308,11 +308,11 @@ font path elements:
 .PP
 .RS 4
 .nf
-.I __projectroot__/lib/X11/fonts/misc/
-.I __projectroot__/lib/X11/fonts/TTF/
-.I __projectroot__/lib/X11/fonts/Type1/
-.I __projectroot__/lib/X11/fonts/75dpi/
-.I __projectroot__/lib/X11/fonts/100dpi/
+.I __projectroot__/share/X11/fonts/misc/
+.I __projectroot__/share/X11/fonts/TTF/
+.I __projectroot__/share/X11/fonts/Type1/
+.I __projectroot__/share/X11/fonts/75dpi/
+.I __projectroot__/share/X11/fonts/100dpi/
 .fi
 .RE
 .PP
@@ -320,14 +320,14 @@ The recommended font path contains the f
 .PP
 .RS 4
 .nf
-.I __projectroot__/lib/X11/fonts/local/
-.I __projectroot__/lib/X11/fonts/misc/
-.I __projectroot__/lib/X11/fonts/75dpi/:unscaled
-.I __projectroot__/lib/X11/fonts/100dpi/:unscaled
-.I __projectroot__/lib/X11/fonts/Type1/
-.I __projectroot__/lib/X11/fonts/Speedo/
-.I __projectroot__/lib/X11/fonts/75dpi/
-.I __projectroot__/lib/X11/fonts/100dpi/
+.I __projectroot__/share/X11/fonts/local/
+.I __projectroot__/share/X11/fonts/misc/
+.I __projectroot__/share/X11/fonts/75dpi/:unscaled
+.I __projectroot__/share/X11/fonts/100dpi/:unscaled
+.I __projectroot__/share/X11/fonts/Type1/
+.I __projectroot__/share/X11/fonts/Speedo/
+.I __projectroot__/share/X11/fonts/75dpi/
+.I __projectroot__/share/X11/fonts/100dpi/
 .fi
 .RE
 .PP
