From: Adam Jackson <ajax@redhat.com>
Date: Sun, 28 Oct 2007 09:37:52 +0100
Subject: [PATCH] Disable pre-release warning.

---
--- xorg-server-1.2.99.905/hw/xfree86/common/xf86Init.c.jx	2007-01-31 19:55:51.000000000 -0500
+++ xorg-server-1.2.99.905/hw/xfree86/common/xf86Init.c	2007-04-11 12:32:24.000000000 -0400
@@ -1770,7 +1770,7 @@
 static void
 xf86PrintBanner()
 {
-#if PRE_RELEASE
+#if 0
   ErrorF("\n"
     "This is a pre-release version of the X server from " XVENDORNAME ".\n"
     "It is not supported in any way.\n"
