From 00fb367cffcc6dd36e5c023757414ed63e13e643 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 3 Mar 2009 14:32:52 -0500
Subject: [PATCH] Don't bother relabeling resources that are being destroyed

---
 Xext/xselinux.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/Xext/xselinux.c b/Xext/xselinux.c
index 6085983..f8495ea 100644
--- a/Xext/xselinux.c
+++ b/Xext/xselinux.c
@@ -1150,6 +1150,8 @@ SELinuxResourceState(CallbackListPtr *pcbl, pointer unused, pointer calldata)
 
     if (rec->type != RT_WINDOW)
 	return;
+    if (rec->state != ResourceStateAdding)
+	return;
 
     pWin = (WindowPtr)rec->value;
     subj = dixLookupPrivate(&wClient(pWin)->devPrivates, subjectKey);
-- 
1.6.1.3

